include/deal.II/lac/la_parallel_vector.templates.h

```diff
@@ -128,27 +129,136 @@ namespace LinearAlgebra
         resize_val(const types::global_dof_index new_alloc_size,
                    types::global_dof_index &     allocated_size,
                    ::dealii::MemorySpace::
-                     MemorySpaceData<Number, ::dealii::MemorySpace::Host> &data)
+                     MemorySpaceData<Number, ::dealii::MemorySpace::Host> &data,
+                   const MPI_Comm &comm_shared)
         {
-          if (new_alloc_size > allocated_size)
+          if (comm_shared == MPI_COMM_SELF)
             {
-              Assert(((allocated_size > 0 && data.values != nullptr) ||
-                      data.values == nullptr),
-                     ExcInternalError());
+              if (new_alloc_size > allocated_size)
+                {
+                  Assert(((allocated_size > 0 && data.values != nullptr) ||
+                          data.values == nullptr),
+                         ExcInternalError());
 
-              Number *new_val;
-              Utilities::System::posix_memalign(
-                reinterpret_cast<void **>(&new_val),
-                64,
-                sizeof(Number) * new_alloc_size);
-              data.values = {new_val, [](Number *data) { std::free(data); }};
+                  Number *new_val;
+                  Utilities::System::posix_memalign(
+                    reinterpret_cast<void **>(&new_val),
+                    64,
+                    sizeof(Number) * new_alloc_size);
+                  data.values = {new_val,
+                                 [](Number *data) { std::free(data); }};
 
-              allocated_size = new_alloc_size;
+                  allocated_size = new_alloc_size;
+                }
+              else if (new_alloc_size == 0)
+                {
+                  data.values.reset();
+                  allocated_size = 0;
+                }
+              data.values_sm = {
+                ArrayView<const Number>(data.values.get(), new_alloc_size)};
             }
-          else if (new_alloc_size == 0)
+          else
             {
-              data.values.reset();
-              allocated_size = 0;
+#ifndef DEAL_II_WITH_MPI
+              Assert(false, ExcInternalError());
+#else
+              allocated_size = new_alloc_size;
+
+              const unsigned int size_sm =
+                Utilities::MPI::n_mpi_processes(comm_shared);
+              const unsigned int rank_sm =
+                Utilities::MPI::this_mpi_process(comm_shared);
+
+              MPI_Win *win = new MPI_Win;
```

Comment:
I must admit I am still unclear about the MPI documentation regarding this function. Due the `MPI_Win_free` later in line 247 (through the deleter of `std::unique_ptr`) also call `free` to release whatever `new` allocated in this line here? Since I find this confusing, others might be confused as well, so you might want to add a comment here.

Reply:
Good point. The delete for the pointer was missing. I have rerun all MatrixFree tests.

Reply:
This is still a bit awkward. Please leave a comment above this line.
We could also make MPI_Win a member of the vector class, so no new/delete is required.

Reply:
I have tried to move the `MPI_Win` to `MemorySpaceData` (this is where it conceptually belongs to) but failed in doing so because I tended to get a segmentation fault. 

I have modified the code now in the following way:
```cpp
MPI_Win mpi_window;

// ...
data.values = {ptr_aligned, [mpi_window](Number *) mutable { MPI_Win_free(&mpi_window);}};
```
Please note that the window is copied. I have added a note.